File: python/rest_api/rest_api/routes_users.py
from fastapi import APIRouter, HTTPException, Query

from .models import PaginatedResponse, User, UserCreate, UserUpdate
from .store import store

router = APIRouter(prefix="/users", tags=["users"])


@router.get("", response_model=PaginatedResponse[User])
def list_users(
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    search: str | None = None,
) -> PaginatedResponse[User]:
    return store.list_users(page, per_page, search)


@router.post("", response_model=User, status_code=201)
def create_user(body: UserCreate) -> User:
    return store.create_user(body)


@router.get("/{user_id}", response_model=User)
def get_user(user_id: int) -> User:
    user = store.get_user(user_id)
    if user is None:
        raise HTTPException(404, "User not found")
    return user


@router.patch("/{user_id}", response_model=User)
def update_user(user_id: int, body: UserUpdate) -> User:
    user = store.update_user(user_id, body)
    if user is None:
        raise HTTPException(404, "User not found")
    return user


@router.delete("/{user_id}", status_code=204)
def delete_user(user_id: int) -> None:
    if not store.delete_user(user_id):
        raise HTTPException(404, "User not found")

File: elixir/rest_api/lib/rest_api/routes_users.ex
defmodule RestApi.RoutesUsers do
  use Plug.Router
  alias RestApi.{Models.User, Store, Router}

  plug(:match)
  plug(:dispatch)

  defp with_post_count(user_json) do
    Map.put(user_json, :post_count, Store.count_user_posts(user_json.id))
  end

  get "/" do
    conn = fetch_query_params(conn)
    page = parse_int(conn.query_params["page"], 1)
    per_page = parse_int(conn.query_params["per_page"], 20)
    search = conn.query_params["search"]
    result = Store.list_users(page, per_page, search)

    result = %{
      result
      | items: Enum.map(result.items, &(&1 |> User.to_json() |> with_post_count()))
    }

    Router.send_json(conn, 200, result)
  end

  post "/" do
    user = Store.create_user(conn.body_params)
    Router.send_json(conn, 201, user |> User.to_json() |> with_post_count())
  end

  get "/:id" do
    case Store.get_user(String.to_integer(id)) do
      nil -> Router.send_json(conn, 404, %{detail: "User not found"})
      user -> Router.send_json(conn, 200, user |> User.to_json() |> with_post_count())
    end
  end

  patch "/:id" do
    case Store.update_user(String.to_integer(id), conn.body_params) do
      nil -> Router.send_json(conn, 404, %{detail: "User not found"})
      user -> Router.send_json(conn, 200, user |> User.to_json() |> with_post_count())
    end
  end

  delete "/:id" do
    case Store.delete_user(String.to_integer(id)) do
      false -> Router.send_json(conn, 404, %{detail: "User not found"})
      true -> send_resp(conn, 204, "")
    end
  end

  defp parse_int(nil, default), do: default
  defp parse_int(val, _default), do: String.to_integer(val)
end
